### PR TITLE
add endpoint for user wallet to claim virtual grant funds

### LINKED ISSUE #173: new claim wallet endpoint that accepts blinded tokens

### PR DESCRIPTION
**Pull request by evq**
currently this endpoint is ready for protocol level integration testing with the browser, it needs one final pass to add the actual funds transfer

depends on https://github.com/brave-intl/bat-ledger/pull/806

fixes #173

### CODE REVIEW

File: promotion/datastore.go [hunk 1]
```diff
@@ -691,3 +695,113 @@ limit 1`
...
+	if err != nil {
+		_ = tx.Rollback()
+		return attempted, err
+	}
+
+	txn, err := worker.RedeemAndTransferFunds(ctx, credentials, job.WalletID, job.Total)
+	if err != nil || txn == nil {
+		// FIXME only non-retriable errors should set erred
+		{
+			_, err := tx.Exec(`update claim_drain set erred = true where id = $1`, job.ID)
```
Comment (maikelmclauflin) [inline promotion/datastore.go]: does this work? it doesn't wipe out all claim queries that came before it?

File: datastore/grantserver/postgres.go [hunk 2]
```diff
@@ -66,7 +66,7 @@ func (pg *Postgres) Migrate() error {
 		return err
 	}
 
-	err = m.Migrate(8)
+	err = m.Migrate(9)
```
Comment (jseidl) [inline datastore/grantserver/postgres.go]: This is the migration version number? Can we use a variable with a meaningful name?
```
MIGRATION_VERSION = 9
m.Migrate(MIGRATION_VERSION)
```

File: promotion/controllers.go [hunk 3]
```diff
@@ -379,6 +380,62 @@ func MakeSuggestion(service *Service) handlers.AppHandler {
...
+
+		err = service.Drain(r.Context(), req.Credentials, req.WalletID)
+		if err != nil {
+			switch err.(type) {
+			case govalidator.Error:
+				return handlers.WrapValidationError(err)
+			case govalidator.Errors:
+				return handlers.WrapValidationError(err)
+			default:
+				// FIXME
```
Comment (jseidl) [inline promotion/controllers.go]: What needs fixing here? Is this related to @maikelmclauflin's comment?

File: promotion/controllers_test.go [hunk 4]
```diff
@@ -899,7 +899,7 @@ func (suite *ControllersTestSuite) TestClaimCompatability() {
 			PublicKey:   hexPublicKey,
 			LastBalance: nil,
 		}
-		suite.Require().NoError(pg.InsertWallet(w), "could not insert wallet")
+		suite.Require().NoError(pg.UpsertWallet(w), "could not insert wallet")
```
Comment (jseidl) [inline promotion/controllers_test.go]: Should this message be `could not upsert wallet` ?

File: promotion/datastore.go [hunk 5]
```diff
@@ -691,3 +695,113 @@ limit 1`
...
+
+// RunNextDrainJob to process deposits if there is one waiting
+func (pg *Postgres) RunNextDrainJob(ctx context.Context, worker DrainWorker) (bool, error) {
+	tx, err := pg.DB.Beginx()
+	attempted := false
+	if err != nil {
+		return attempted, err
+	}
+
+	// FIXME
```
Comment (jseidl) [inline promotion/datastore.go]: What needs fixing here?

File: promotion/datastore_test.go [hunk 6]
```diff
@@ -739,3 +737,93 @@ func (suite *PostgresTestSuite) TestInsertClobberedClaims() {
 	err = pg.DB.Select(&allCreds2, `select * from clobbered_claims;`)
 	suite.Require().Equal(allCreds1, allCreds2, "creds should not be inserted more than once")
 }
+
+func (suite *PostgresTestSuite) TestDrainClaim() {
+	pg, err := NewPostgres("", false)
+	suite.Require().NoError(err)
+
+	publicKey := "hBrtClwIppLmu/qZ8EhGM1TQZUwDUosbOrVu3jMwryY="
+	blindedCreds := jsonutils.JSONStringArray([]string{"hBrtClwIppLmu/qZ8EhGM1TQZUwDUosbOrVu3jMwryY="})
```
Comment (jseidl) [inline promotion/datastore_test.go]: Shouldn't this be `publicKey` instead of the same string from above?

File: promotion/datastore_test.go [hunk 7]
```diff
@@ -739,3 +737,93 @@ func (suite *PostgresTestSuite) TestInsertClobberedClaims() {
...
+			attempted, err = pg.RunNextClaimJob(context.Background(), mockClaimWorker)
+			suite.Require().Equal(true, attempted)
+			suite.Require().NoError(err)
+
+			// No further jobs should run after success
+			attempted, err = pg.RunNextClaimJob(context.Background(), mockClaimWorker)
+			suite.Require().Equal(false, attempted)
+			suite.Require().NoError(err)
+
+	*/
```
Comment (jseidl) [inline promotion/datastore_test.go]: Can we remove this commented-out code?

File: promotion/drain.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+	// Iterate through each credential and assemble list of funding sources
+	_, _, fundingSources, err := service.GetCredentialRedemptions(ctx, credentials)
+	if err != nil {
+		return err
+	}
+
+	for _, v := range fundingSources {
+		if v.Type != "ads" {
+			return errors.New("Only ads suggestions can be drained")
```
Comment (jseidl) [inline promotion/drain.go]: Since this is a loop, should invalid sources be skipped rather than failing everything?

File: promotion/drain.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+	if wallet == nil || wallet.PayoutAddress == nil {
+		return nil, errors.New("missing wallet")
+	}
+
+	err = service.cbClient.RedeemCredentials(ctx, credentials, walletID.String())
+	if err != nil {
+		return nil, err
+	}
+
+	// FIXME
```
Comment (jseidl) [inline promotion/drain.go]: What needs fixing here?

File: promotion/controllers_test.go (hunk 4 above)
Comment (evq) [reply promotion/controllers_test.go]: the upsert is just inserting since our dbs will be clean at the start of the test

File: promotion/datastore_test.go (hunk 6 above)
Comment (evq) [reply promotion/datastore_test.go]: technically it should be a distinct value in each - but here since we're mocking the calls I just reused it. so it may be more confusing to reference `publicKey`

File: promotion/drain.go (hunk 8 above)
Comment (evq) [reply promotion/drain.go]: I think that is less safe then aborting. If in such a case we return success we run the risk of the client proceeding to delete tokens that the server was in fact unable to process ( unbeknownst to the client ). We could consider tracking and returning information about which funding sources were valid and able to be processed but then we're complicating the API and adding more surface area for issues.

File: promotion/controllers.go [hunk 10]
```diff
@@ -379,6 +380,62 @@ func MakeSuggestion(service *Service) handlers.AppHandler {
...
+		}
+
+		logging.AddWalletIDToContext(r.Context(), req.WalletID)
+
+		keyID, err := middleware.GetKeyID(r.Context())
+		if err != nil {
+			return handlers.WrapError(err, "Error looking up http signature info", http.StatusBadRequest)
+		}
+		if req.WalletID.String() != keyID {
+			return &handlers.AppError{
```
[the remark this answers is not in the document]
Comment (evq) [reply promotion/controllers.go]: fixed and opened https://github.com/brave-intl/bat-go/issues/266 for us to switch everything over to use it

**Comment by evq** [comment]
i'm going to merge this into master so it get's deployed to dev
